### PR TITLE
Added nn.Module support for FasterRCNN backbone

### LINKED ISSUE #660: Add ability add custom/modified backbone to FasterRCNN
## 🚀 Feature
In the [Faster RCNN](https://github.com/PyTorchLightning/lightning-bolts/blob/master/pl_bolts/models/detection/faster_rcnn/faster_rcnn_module.py) bolt, it would be awesome to have the functionality to pass custom `pl.LightningModules` or `torch.nn.Module` backbone. In the current state, the constructor requires the `backbone` argument to be a `str`

https://github.com/PyTorchLightning/lightning-bolts/blob/c3b60de7dc30c5f7947256479d9be3a042b8c182/pl_bolts/models/detection/faster_rcnn/faster_rcnn_module.py#L33-L63

### Motivation

I wanted to try some custom version of the ResNet model in the Faster RCNN pipeline and thought this functionality might help.

### Alternatives

These are the two workarounds that I think would work:
1. Reassign the backbone manually. Similar to how we reassign the predictor head in torchvision tutorials - [link](https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html#an-instance-segmentation-model-for-pennfudan-dataset). 
2. Subclass the `FasterRCNN` and manually reassign the backbone

### PR DESCRIPTION
**Pull request by abhayraw1**
## What does this PR do?

Fixes #660 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Sure! 🙃

### CODE REVIEW

**Comment by abhayraw1** [comment]
@Borda, I made the changes. Let me know what you think!

PS: Tests had failed due to the issue below, which I see has been fixed now...

> This change is causing tests to fail in master. Please check!
> 
> [3bf65da#diff-81b21ad681f00e8e80279aaf4d5e2c3a304b5a587b79850cff8d4472147f7f0aL92](https://github.com/PyTorchLightning/lightning-bolts/commit/3bf65da4d611445a9b09399f7bcee68510c8ca13#diff-81b21ad681f00e8e80279aaf4d5e2c3a304b5a587b79850cff8d4472147f7f0aL92)

File: pl_bolts/models/detection/faster_rcnn/faster_rcnn_module.py [hunk 1]
```diff
@@ -90,13 +90,22 @@ def __init__(
...
-            backbone_model = create_fasterrcnn_backbone(
-                self.backbone,
-                fpn,
-                pretrained_backbone,
-                trainable_backbone_layers,
-                **kwargs,
-            )
+            if isinstance(self.backbone, torch.nn.Module):
+                backbone_model = self.backbone
+                if pretrained_backbone:
```
Comment (Borda) [inline pl_bolts/models/detection/faster_rcnn/faster_rcnn_module.py:95]: shall we add some check on the required dimensionality, I believe that you cannot add any backbone, right?

Comment (Borda) [reply pl_bolts/models/detection/faster_rcnn/faster_rcnn_module.py:95]: @abhayraw1 ^^

**Review by Borda — APPROVED** [review]

File: pl_bolts/models/detection/faster_rcnn/faster_rcnn_module.py (hunk 1 above)
Comment (SeanNaren) [reply pl_bolts/models/detection/faster_rcnn/faster_rcnn_module.py:91]: To unblock this PR, I think this is fine for now :)

**Review by SeanNaren — APPROVED** [review]

**Comment by SeanNaren** [comment]
Hey @abhayraw1 seems there is one failing test left, do you mind checking it out! thanks for your work so far!!

**Review by akihironitta — APPROVED** [review]
@abhayraw1 LGTM!

**Comment by Borda** [comment]
Not sure what is wrong here:
```
                boxes = target["boxes"]
                degenerate_boxes = boxes[:, 2:] <= boxes[:, :2]
                if degenerate_boxes.any():
                    # print the first degenerate box
                    bb_idx = torch.where(degenerate_boxes.any(dim=1))[0][0]
                    degen_bb: List[float] = boxes[bb_idx].tolist()
>                   raise ValueError("All bounding boxes should have positive height and width."
                                     " Found invalid box {} for target at index {}."
                                     .format(degen_bb, target_idx))
E                   ValueError: All bounding boxes should have positive height and width. Found invalid box [137.5, 556.25, 743.75, 556.25] for target at index 0.
```
as I see the `[137.5, 556.25] <=  [743.75, 556.25]` shall be true, but it fails...

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Pull request merged**